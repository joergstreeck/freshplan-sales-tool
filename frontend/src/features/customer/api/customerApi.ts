import { httpClient } from '../../../lib/apiClient';
import type { CustomerListResponse } from '../types/customer.types';

export const customerApi = {
  // Get paginated list of customers
  getCustomers: async (
    page = 0,
    size = 20,
    sortBy = 'companyName'
  ): Promise<CustomerListResponse> => {
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
      sort: sortBy,
    });
    const response = await httpClient.get<CustomerListResponse>(`/api/customers?${params}`);
    return response.data;
  },

  // Get customers by status
  getCustomersByStatus: async (
    status: string,
    page = 0,
    size = 20
  ): Promise<CustomerListResponse> => {
    const params = new URLSearchParams({
      status,
      page: page.toString(),
      size: size.toString(),
    });
    const response = await httpClient.get<CustomerListResponse>(`/api/customers?${params}`);
    return response.data;
  },

  // Search customers
  searchCustomers: async (query: string, page = 0, size = 20): Promise<CustomerListResponse> => {
    const params = new URLSearchParams({
      q: query,
      page: page.toString(),
      size: size.toString(),
    });
    const response = await httpClient.get<CustomerListResponse>(`/api/customers/search?${params}`);
    return response.data;
  },
};
