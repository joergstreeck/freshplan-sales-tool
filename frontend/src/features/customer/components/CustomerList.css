/* CustomerList Component Styles - FreshPlan CI */

.customer-list-container {
  width: 100%;
  padding: var(--spacing-xl);
  background-color: var(--color-background);
  min-height: 100vh;
}

/* Header */
.customer-list-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: var(--spacing-xl);
  padding-bottom: var(--spacing-lg);
  border-bottom: 2px solid var(--color-primary);
}

.customer-list-header h1 {
  font-family: var(--font-heading);
  font-size: var(--font-size-3xl);
  color: var(--color-secondary);
  margin: 0;
  font-weight: var(--font-weight-bold);
}

.customer-list-stats {
  display: flex;
  gap: var(--spacing-lg);
}

.stat-item {
  font-family: var(--font-body);
  font-size: var(--font-size-base);
  color: var(--color-text);
}

.stat-item strong {
  color: var(--color-primary);
  font-size: var(--font-size-xl);
  margin-right: var(--spacing-xs);
}

/* Table Wrapper */
.customer-list-table-wrapper {
  background: var(--color-white);
  border-radius: var(--radius-lg);
  box-shadow: var(--shadow-lg);
  overflow: hidden;
  margin-bottom: var(--spacing-xl);
}

/* Table */
.customer-list-table {
  width: 100%;
  border-collapse: collapse;
  font-family: var(--font-body);
}

.customer-list-table thead {
  background: var(--bg-gradient-primary);
  color: var(--color-white);
}

.customer-list-table th {
  padding: var(--spacing-md) var(--spacing-lg);
  text-align: left;
  font-weight: var(--font-weight-medium);
  font-size: var(--font-size-sm);
  text-transform: uppercase;
  letter-spacing: 0.5px;
  white-space: nowrap;
  position: relative;
}

.customer-list-table th.sortable {
  cursor: pointer;
  user-select: none;
  transition: opacity var(--transition-fast);
}

.customer-list-table th.sortable:hover {
  opacity: 0.8;
}

.customer-list-table th.sortable.sorted::after {
  content: ' â†“';
  position: absolute;
  right: var(--spacing-sm);
}

.customer-list-table tbody tr {
  border-bottom: 1px solid var(--border-light);
  transition: background-color var(--transition-fast);
}

.customer-list-table tbody tr:hover {
  background-color: var(--color-primary-light);
}

.customer-list-table tbody tr.at-risk {
  background-color: rgba(244, 67, 54, 0.05);
}

.customer-list-table tbody tr.at-risk:hover {
  background-color: rgba(244, 67, 54, 0.1);
}

.customer-list-table td {
  padding: var(--spacing-md) var(--spacing-lg);
  font-size: var(--font-size-sm);
  color: var(--color-text);
}

/* Specific Column Styles */
.customer-number {
  font-family: monospace;
  color: var(--color-text-light);
  font-size: var(--font-size-xs);
}

.company-name {
  display: flex;
  flex-direction: column;
  gap: var(--spacing-xs);
}

.company-name strong {
  color: var(--color-secondary);
  font-weight: var(--font-weight-medium);
}

.trading-name {
  font-size: var(--font-size-xs);
  color: var(--color-text-light);
  font-style: italic;
}

.currency {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

/* Status Badge */
.status-badge {
  display: inline-block;
  padding: var(--spacing-xs) var(--spacing-sm);
  border-radius: var(--radius-full);
  font-size: var(--font-size-xs);
  font-weight: var(--font-weight-medium);
  color: var(--color-white);
  text-transform: uppercase;
  letter-spacing: 0.5px;
}

/* Risk Indicator */
.risk-indicator {
  display: flex;
  align-items: center;
  gap: var(--spacing-sm);
  min-width: 120px;
}

.risk-bar {
  height: 6px;
  background-color: #4CAF50;
  border-radius: var(--radius-sm);
  transition: width var(--transition-medium);
  flex: 1;
  max-width: 60px;
}

.risk-score {
  font-size: var(--font-size-xs);
  color: var(--color-text-light);
  font-weight: var(--font-weight-medium);
  min-width: 35px;
  text-align: right;
}

/* Pagination */
.customer-list-pagination {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: var(--spacing-lg);
  background: var(--color-white);
  border-radius: var(--radius-lg);
  box-shadow: var(--shadow-md);
}

.pagination-info {
  font-size: var(--font-size-sm);
  color: var(--color-text-light);
}

.pagination-controls {
  display: flex;
  align-items: center;
  gap: var(--spacing-sm);
}

.pagination-button {
  padding: var(--spacing-sm) var(--spacing-md);
  background: var(--color-primary);
  color: var(--color-white);
  border: none;
  border-radius: var(--radius-md);
  font-family: var(--font-body);
  font-size: var(--font-size-sm);
  font-weight: var(--font-weight-medium);
  cursor: pointer;
  transition: all var(--transition-fast);
}

.pagination-button:hover:not(:disabled) {
  background: var(--color-primary-dark);
  transform: translateY(-1px);
  box-shadow: var(--shadow-md);
}

.pagination-button:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.pagination-current {
  padding: 0 var(--spacing-lg);
  font-size: var(--font-size-sm);
  color: var(--color-text);
  font-weight: var(--font-weight-medium);
}

/* Loading State */
.customer-list-loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 400px;
  gap: var(--spacing-lg);
}

.customer-list-loading p {
  font-size: var(--font-size-lg);
  color: var(--color-text-light);
}

/* Error State */
.customer-list-error {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 400px;
  text-align: center;
  padding: var(--spacing-xl);
}

.error-icon {
  font-size: 4rem;
  margin-bottom: var(--spacing-lg);
}

.customer-list-error h3 {
  font-family: var(--font-heading);
  font-size: var(--font-size-2xl);
  color: var(--danger);
  margin-bottom: var(--spacing-md);
}

.customer-list-error p {
  font-size: var(--font-size-base);
  color: var(--color-text-light);
  max-width: 500px;
}

/* Empty State */
.customer-list-empty {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 400px;
  text-align: center;
  padding: var(--spacing-xl);
}

.empty-icon {
  font-size: 4rem;
  margin-bottom: var(--spacing-lg);
  opacity: 0.5;
}

.customer-list-empty h3 {
  font-family: var(--font-heading);
  font-size: var(--font-size-2xl);
  color: var(--color-text);
  margin-bottom: var(--spacing-md);
}

.customer-list-empty p {
  font-size: var(--font-size-base);
  color: var(--color-text-light);
}

/* Responsive Design */
@media (max-width: 1200px) {
  .customer-list-container {
    padding: var(--spacing-lg);
  }
  
  .customer-list-table {
    font-size: var(--font-size-sm);
  }
  
  .customer-list-table th,
  .customer-list-table td {
    padding: var(--spacing-sm) var(--spacing-md);
  }
}

@media (max-width: 768px) {
  .customer-list-header {
    flex-direction: column;
    align-items: flex-start;
    gap: var(--spacing-md);
  }
  
  .customer-list-table-wrapper {
    overflow-x: auto;
  }
  
  .customer-list-table {
    min-width: 800px;
  }
  
  .customer-list-pagination {
    flex-direction: column;
    gap: var(--spacing-md);
  }
  
  .pagination-controls {
    flex-wrap: wrap;
    justify-content: center;
  }
}