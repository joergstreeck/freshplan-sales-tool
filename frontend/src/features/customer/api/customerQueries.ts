import { useQuery } from '@tanstack/react-query';
import { customerApi } from './customerApi';
import type { CustomerStatus } from '../types/customer.types';

// Query Keys
export const customerKeys = {
  all: ['customers'] as const,
  lists: () => [...customerKeys.all, 'list'] as const,
  list: (page?: number, size?: number, sortBy?: string) =>
    [...customerKeys.lists(), { page, size, sortBy }] as const,
  byStatus: (status: CustomerStatus, page?: number, size?: number) =>
    [...customerKeys.lists(), 'status', status, { page, size }] as const,
  search: (query: string, page?: number, size?: number) =>
    [...customerKeys.all, 'search', query, { page, size }] as const,
};

// Hooks
export const useCustomers = (page = 0, size = 20, sortBy = 'companyName') => {
  return useQuery({
    queryKey: customerKeys.list(page, size, sortBy),
    queryFn: () => customerApi.getCustomers(page, size, sortBy),
    staleTime: 1000 * 60 * 5, // 5 minutes
  });
};

export const useCustomersByStatus = (
  status: CustomerStatus,
  page = 0,
  size = 20,
  enabled = true
) => {
  return useQuery({
    queryKey: customerKeys.byStatus(status, page, size),
    queryFn: () => customerApi.getCustomersByStatus(status, page, size),
    enabled,
    staleTime: 1000 * 60 * 5,
  });
};

export const useCustomerSearch = (query: string, page = 0, size = 20, enabled = true) => {
  return useQuery({
    queryKey: customerKeys.search(query, page, size),
    queryFn: () => customerApi.searchCustomers(query, page, size),
    enabled: enabled && query.length > 0,
    staleTime: 1000 * 60 * 2, // 2 minutes for search results
  });
};

//Server-Side Filtering with advanced filters
export const useCustomerSearchAdvanced = (
  searchRequest: {
    globalSearch?: string;
    filters?: Array<{
      field: string;
      operator: string;
      value: string | string[];
    }>;
    sort?: {
      field: string;
      direction: 'ASC' | 'DESC';
    };
  },
  page = 0,
  size = 50,
  enabled = true
) => {
  return useQuery({
    queryKey: [...customerKeys.all, 'advanced-search', searchRequest, page, size],
    queryFn: () => customerApi.searchCustomersAdvanced(searchRequest, page, size),
    enabled,
    staleTime: 1000 * 30, // 30 seconds
    gcTime: 1000 * 60 * 5, // 5 minutes
  });
};
