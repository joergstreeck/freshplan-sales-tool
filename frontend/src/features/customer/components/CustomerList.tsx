import React, { useState } from 'react';
import { useCustomers } from '../api/customerQueries';
import type { CustomerResponse } from '../types/customer.types';
import {
  customerTypeLabels,
  customerStatusLabels,
  industryLabels,
  customerStatusColors,
} from '../types/customer.types';
import { LoadingSpinner } from '../../../components/ui/LoadingSpinner';
import { UniversalExportButton } from '../../../components/export';
import './CustomerList.css';

export const CustomerList: React.FC = () => {
  const [currentPage, setCurrentPage] = useState(0);
  const [sortBy, setSortBy] = useState('companyName');
  const pageSize = 50; // Erh√∂ht von 20 auf 50 f√ºr bessere √úbersicht

  const { data, isLoading, isError, error } = useCustomers(currentPage, pageSize, sortBy);

  // Format currency
  const formatCurrency = (amount?: number) => {
    if (!amount) return '-';
    return new Intl.NumberFormat('de-DE', {
      style: 'currency',
      currency: 'EUR',
    }).format(amount);
  };

  // Format date
  const formatDate = (dateString?: string) => {
    if (!dateString) return '-';
    return new Date(dateString).toLocaleDateString('de-DE', {
      day: '2-digit',
      month: '2-digit',
      year: 'numeric',
    });
  };

  // Handle sorting
  const handleSort = (field: string) => {
    setSortBy(field);
    setCurrentPage(0); // Reset to first page when sorting
  };

  // Render loading state
  if (isLoading) {
    return (
      <div className="customer-list-loading">
        <LoadingSpinner />
        <p>Kundendaten werden geladen...</p>
      </div>
    );
  }

  // Render error state
  if (isError) {
    return (
      <div className="customer-list-error">
        <div className="error-icon">‚ö†Ô∏è</div>
        <h3>Fehler beim Laden der Kundendaten</h3>
        <p>{error instanceof Error ? error.message : 'Ein unerwarteter Fehler ist aufgetreten.'}</p>
      </div>
    );
  }

  // Render empty state
  if (!data?.content || data.content.length === 0) {
    return (
      <div className="customer-list-empty">
        <div className="empty-icon">üìã</div>
        <h3>Keine Kunden gefunden</h3>
        <p>Es sind noch keine Kunden im System vorhanden.</p>
      </div>
    );
  }

  return (
    <div className="customer-list-container">
      {/* Header */}
      <div className="customer-list-header">
        <h1>Kundenliste</h1>
        <div className="customer-list-actions">
          <div className="customer-list-stats">
            <span className="stat-item">
              <strong>{data.totalElements}</strong> Kunden gesamt
            </span>
          </div>
          <UniversalExportButton
            entity="customers"
            buttonLabel="Liste exportieren"
            onExportComplete={format => {}}
          />
          <a href="/kundenmanagement/neu" className="btn-new-customer">
            + Neuer Kunde
          </a>
        </div>
      </div>

      {/* Table */}
      <div className="customer-list-table-wrapper">
        <table className="customer-list-table">
          <thead>
            <tr>
              <th
                className={`sortable ${sortBy === 'customerNumber' ? 'sorted' : ''}`}
                onClick={() => handleSort('customerNumber')}
              >
                Kundennummer
              </th>
              <th
                className={`sortable ${sortBy === 'companyName' ? 'sorted' : ''}`}
                onClick={() => handleSort('companyName')}
              >
                Firmenname
              </th>
              <th>Typ</th>
              <th>Branche</th>
              <th
                className={`sortable ${sortBy === 'status' ? 'sorted' : ''}`}
                onClick={() => handleSort('status')}
              >
                Status
              </th>
              <th>Jahresumsatz</th>
              <th
                className={`sortable ${sortBy === 'lastContactDate' ? 'sorted' : ''}`}
                onClick={() => handleSort('lastContactDate')}
              >
                Letzter Kontakt
              </th>
              <th>Risiko</th>
            </tr>
          </thead>
          <tbody>
            {data.content.map((customer: CustomerResponse) => (
              <tr key={customer.id} className={customer.atRisk ? 'at-risk' : ''}>
                <td className="customer-number">{customer.customerNumber}</td>
                <td className="company-name">
                  <strong>{customer.companyName}</strong>
                  {customer.tradingName && (
                    <span className="trading-name">{customer.tradingName}</span>
                  )}
                </td>
                <td>{customerTypeLabels[customer.customerType]}</td>
                <td>{customer.industry ? industryLabels[customer.industry] : '-'}</td>
                <td>
                  <span
                    className="status-badge"
                    style={{ backgroundColor: customerStatusColors[customer.status] }}
                  >
                    {customerStatusLabels[customer.status]}
                  </span>
                </td>
                <td className="currency">{formatCurrency(customer.expectedAnnualVolume)}</td>
                <td>{formatDate(customer.lastContactDate)}</td>
                <td>
                  <div className="risk-indicator">
                    <div
                      className="risk-bar"
                      style={{
                        width: `${customer.riskScore}%`,
                        backgroundColor:
                          customer.riskScore > 70
                            ? '#F44336'
                            : customer.riskScore > 40
                              ? '#FF9800'
                              : '#4CAF50',
                      }}
                    />
                    <span className="risk-score">{customer.riskScore}%</span>
                  </div>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>

      {/* Pagination */}
      <div className="customer-list-pagination">
        <div className="pagination-info">
          Zeige {data.page * data.size + 1} -{' '}
          {Math.min((data.page + 1) * data.size, data.totalElements)} von {data.totalElements}{' '}
          Eintr√§gen
        </div>
        <div className="pagination-controls">
          <button
            className="pagination-button"
            onClick={() => setCurrentPage(0)}
            disabled={data.first}
          >
            Erste
          </button>
          <button
            className="pagination-button"
            onClick={() => setCurrentPage(prev => Math.max(0, prev - 1))}
            disabled={data.first}
          >
            Zur√ºck
          </button>
          <span className="pagination-current">
            Seite {data.page + 1} von {data.totalPages}
          </span>
          <button
            className="pagination-button"
            onClick={() => setCurrentPage(prev => prev + 1)}
            disabled={data.last}
          >
            Weiter
          </button>
          <button
            className="pagination-button"
            onClick={() => setCurrentPage(data.totalPages - 1)}
            disabled={data.last}
          >
            Letzte
          </button>
        </div>
      </div>
    </div>
  );
};
