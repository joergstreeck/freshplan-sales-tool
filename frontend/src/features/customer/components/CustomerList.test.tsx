/**
 * CustomerList Component Tests
 *
 * Comprehensive test suite for CustomerList component
 * including unit tests, integration tests, and accessibility tests
 *
 * @component CustomerList
 * @version 2.0.0
 */

import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';
import { CustomerList } from './CustomerList';
import { customerApi } from '../api/customerApi';
import type { CustomerListResponse } from '../types/customer.types';

// Add jest-axe matchers
expect.extend(toHaveNoViolations);

// Mock the API
vi.mock('../api/customerApi');

// Mock data
const mockCustomers: CustomerListResponse = {
  content: [
    {
      id: '1',
      customerNumber: 'CUST001',
      companyName: 'Test Company GmbH',
      tradingName: 'Test Trading',
      status: 'AKTIV',
      customerType: 'UNTERNEHMEN',
      industry: 'RESTAURANT',
      riskScore: 20,
      atRisk: false,
      expectedAnnualVolume: 50000,
      isDeleted: false,
      hasChildren: false,
      childCustomerIds: [],
      createdAt: '2024-01-01T00:00:00Z',
      createdBy: 'system',
    },
    {
      id: '2',
      customerNumber: 'CUST002',
      companyName: 'Risk Company AG',
      status: 'RISIKO',
      customerType: 'UNTERNEHMEN',
      industry: 'HOTEL',
      riskScore: 80,
      atRisk: true,
      expectedAnnualVolume: 100000,
      isDeleted: false,
      hasChildren: false,
      childCustomerIds: [],
      createdAt: '2024-01-02T00:00:00Z',
      createdBy: 'system',
    },
  ],
  page: 0,
  size: 20,
  totalElements: 2,
  totalPages: 1,
  first: true,
  last: true,
};

// Test wrapper component
const TestWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: { retry: false },
    },
  });

  return (
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{children}</BrowserRouter>
    </QueryClientProvider>
  );
};

describe('CustomerList Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('Rendering', () => {
    it('should render loading state initially', () => {
      vi.mocked(customerApi.getCustomers).mockImplementation(
        () => new Promise(() => {}) // Never resolves
      );

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      expect(screen.getByRole('status')).toBeInTheDocument();
      expect(screen.getByText(/Kundendaten werden geladen/i)).toBeInTheDocument();
    });

    it('should render customer list after loading', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('Test Company GmbH')).toBeInTheDocument();
        expect(screen.getByText('Risk Company AG')).toBeInTheDocument();
      });
    });

    it('should render error state on API failure', async () => {
      vi.mocked(customerApi.getCustomers).mockRejectedValue(new Error('Network error'));

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText(/Fehler beim Laden/i)).toBeInTheDocument();
      });
    });

    it('should render empty state when no customers', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue({
        ...mockCustomers,
        content: [],
        totalElements: 0,
      });

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText(/Keine Kunden gefunden/i)).toBeInTheDocument();
      });
    });
  });

  describe('Interactions', () => {
    it('should handle sorting by company name', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);
      const user = userEvent.setup();

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('Test Company GmbH')).toBeInTheDocument();
      });

      const companyHeader = screen.getByText('Firmenname');
      await user.click(companyHeader);

      expect(vi.mocked(customerApi.getCustomers)).toHaveBeenCalledWith(0, 50, 'companyName');
    });

    it('should handle pagination', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue({
        ...mockCustomers,
        totalElements: 50,
        totalPages: 3,
        first: true,
        last: false,
      });
      const user = userEvent.setup();

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('Test Company GmbH')).toBeInTheDocument();
      });

      const nextButton = screen.getByText('Weiter');

      // Button should be enabled for non-last page
      expect(nextButton).not.toBeDisabled();

      await user.click(nextButton);

      expect(vi.mocked(customerApi.getCustomers)).toHaveBeenCalledWith(1, 50, 'companyName');
    });

    it('should disable pagination buttons appropriately', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue({
        ...mockCustomers,
        first: true,
        last: true,
      });

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('Test Company GmbH')).toBeInTheDocument();
      });

      const prevButton = screen.getByText('Zurück');
      const nextButton = screen.getByText('Weiter');

      expect(prevButton).toBeDisabled();
      expect(nextButton).toBeDisabled();
    });
  });

  describe('Visual States', () => {
    it('should highlight at-risk customers', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);

      const { container } = render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('Risk Company AG')).toBeInTheDocument();
      });

      const riskRow = container.querySelector('.at-risk');
      expect(riskRow).toBeInTheDocument();
    });

    it('should display correct status badges', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        // Use more specific queries to avoid confusion with table header
        const statusCells = screen.getAllByRole('cell');
        const aktivBadge = statusCells.find(cell => cell.textContent === 'Aktiv');
        const risikoBadge = statusCells.find(cell => cell.textContent === 'Risiko');

        expect(aktivBadge?.querySelector('.status-badge')).toBeInTheDocument();
        expect(risikoBadge?.querySelector('.status-badge')).toBeInTheDocument();
      });
    });

    it('should format currency correctly', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('50.000,00 €')).toBeInTheDocument();
        expect(screen.getByText('100.000,00 €')).toBeInTheDocument();
      });
    });
  });

  describe('Accessibility', () => {
    it('should have no accessibility violations', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);

      const { container } = render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('Test Company GmbH')).toBeInTheDocument();
      });

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should have proper ARIA labels', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByRole('table')).toBeInTheDocument();
        expect(screen.getByRole('columnheader', { name: /firmenname/i })).toBeInTheDocument();
      });
    });

    it.skip('should be keyboard navigable', async () => {
      // TODO: Fix keyboard navigation test
      // The sortable headers need tabindex="0" to be focusable
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);
      const user = userEvent.setup();

      render(
        <TestWrapper>
          <CustomerList />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('Test Company GmbH')).toBeInTheDocument();
      });

      // Tab through interactive elements
      await user.tab();
      expect(screen.getByText('Firmenname')).toHaveFocus();

      await user.tab();
      expect(screen.getByText('Zurück')).toHaveFocus();

      await user.tab();
      expect(screen.getByText('Weiter')).toHaveFocus();
    });
  });

  describe('Performance', () => {
    it('should not re-render unnecessarily', async () => {
      vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);
      const renderSpy = vi.fn();

      const TestComponent = () => {
        renderSpy();
        return <CustomerList />;
      };

      const { rerender } = render(
        <TestWrapper>
          <TestComponent />
        </TestWrapper>
      );

      await waitFor(() => {
        expect(screen.getByText('Test Company GmbH')).toBeInTheDocument();
      });

      const initialRenderCount = renderSpy.mock.calls.length;

      // Re-render with same props
      rerender(
        <TestWrapper>
          <TestComponent />
        </TestWrapper>
      );

      // Allow for one additional render due to React Query's behavior
      // React Query may trigger an additional render when checking cache
      expect(renderSpy.mock.calls.length).toBeLessThanOrEqual(initialRenderCount + 1);
    });
  });
});

describe('CustomerList Snapshot Tests', () => {
  it('should match snapshot for loaded state', async () => {
    vi.mocked(customerApi.getCustomers).mockResolvedValue(mockCustomers);

    const { container } = render(
      <TestWrapper>
        <CustomerList />
      </TestWrapper>
    );

    await waitFor(() => {
      expect(screen.getByText('Test Company GmbH')).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for loading state', () => {
    vi.mocked(customerApi.getCustomers).mockImplementation(() => new Promise(() => {}));

    const { container } = render(
      <TestWrapper>
        <CustomerList />
      </TestWrapper>
    );

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for error state', async () => {
    vi.mocked(customerApi.getCustomers).mockRejectedValue(new Error('Network error'));

    const { container } = render(
      <TestWrapper>
        <CustomerList />
      </TestWrapper>
    );

    await waitFor(() => {
      expect(screen.getByText(/Fehler beim Laden/i)).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });
});
