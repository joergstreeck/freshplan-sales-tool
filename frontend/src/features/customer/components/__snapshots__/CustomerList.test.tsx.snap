// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomerList Snapshot Tests > should match snapshot for error state 1`] = `
<div>
  <div
    class="customer-list-error"
  >
    <div
      class="error-icon"
    >
      ⚠️
    </div>
    <h3>
      Fehler beim Laden der Kundendaten
    </h3>
    <p>
      Network error
    </p>
  </div>
</div>
`;

exports[`CustomerList Snapshot Tests > should match snapshot for loaded state 1`] = `
<div>
  <div
    class="customer-list-container"
  >
    <div
      class="customer-list-header"
    >
      <h1>
        Kundenliste
      </h1>
      <div
        class="customer-list-actions"
      >
        <div
          class="customer-list-stats"
        >
          <span
            class="stat-item"
          >
            <strong>
              2
            </strong>
             Kunden gesamt
          </span>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-c9ai1l-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1sh91j5-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="DownloadIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5 20h14v-2H5zM19 9h-4V3H9v6H5l7 7z"
              />
            </svg>
          </span>
          Liste exportieren
        </button>
        <a
          class="btn-new-customer"
          href="/kundenmanagement/neu"
        >
          + Neuer Kunde
        </a>
      </div>
    </div>
    <div
      class="customer-list-table-wrapper"
    >
      <table
        class="customer-list-table"
      >
        <thead>
          <tr>
            <th
              class="sortable "
            >
              Kundennummer
            </th>
            <th
              class="sortable sorted"
            >
              Firmenname
            </th>
            <th>
              Typ
            </th>
            <th>
              Branche
            </th>
            <th
              class="sortable "
            >
              Status
            </th>
            <th>
              Jahresumsatz
            </th>
            <th
              class="sortable "
            >
              Letzter Kontakt
            </th>
            <th>
              Risiko
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
          >
            <td
              class="customer-number"
            >
              CUST001
            </td>
            <td
              class="company-name"
            >
              <strong>
                Test Company GmbH
              </strong>
              <span
                class="trading-name"
              >
                Test Trading
              </span>
            </td>
            <td>
              Unternehmen
            </td>
            <td>
              Restaurant
            </td>
            <td>
              <span
                class="status-badge"
                style="background-color: rgb(76, 175, 80);"
              >
                Aktiv
              </span>
            </td>
            <td
              class="currency"
            >
              50.000,00 €
            </td>
            <td>
              -
            </td>
            <td>
              <div
                class="risk-indicator"
              >
                <div
                  class="risk-bar"
                  style="width: 20%; background-color: rgb(76, 175, 80);"
                />
                <span
                  class="risk-score"
                >
                  20
                  %
                </span>
              </div>
            </td>
          </tr>
          <tr
            class="at-risk"
          >
            <td
              class="customer-number"
            >
              CUST002
            </td>
            <td
              class="company-name"
            >
              <strong>
                Risk Company AG
              </strong>
            </td>
            <td>
              Unternehmen
            </td>
            <td>
              Hotel
            </td>
            <td>
              <span
                class="status-badge"
                style="background-color: rgb(244, 67, 54);"
              >
                Risiko
              </span>
            </td>
            <td
              class="currency"
            >
              100.000,00 €
            </td>
            <td>
              -
            </td>
            <td>
              <div
                class="risk-indicator"
              >
                <div
                  class="risk-bar"
                  style="width: 80%; background-color: rgb(244, 67, 54);"
                />
                <span
                  class="risk-score"
                >
                  80
                  %
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="customer-list-pagination"
    >
      <div
        class="pagination-info"
      >
        Zeige 
        1
         -
         
        2
         von 
        2
         
        Einträgen
      </div>
      <div
        class="pagination-controls"
      >
        <button
          class="pagination-button"
          disabled=""
        >
          Erste
        </button>
        <button
          class="pagination-button"
          disabled=""
        >
          Zurück
        </button>
        <span
          class="pagination-current"
        >
          Seite 
          1
           von 
          1
        </span>
        <button
          class="pagination-button"
          disabled=""
        >
          Weiter
        </button>
        <button
          class="pagination-button"
          disabled=""
        >
          Letzte
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`CustomerList Snapshot Tests > should match snapshot for loading state 1`] = `
<div>
  <div
    class="customer-list-loading"
  >
    <div
      class="flex flex-col items-center justify-center p-4"
    >
      <div
        aria-label="Loading"
        class="w-8 h-8 border-4 border-gray-200 
          border-t-blue-600 rounded-full animate-spin"
        role="status"
      />
    </div>
    <p>
      Kundendaten werden geladen...
    </p>
  </div>
</div>
`;
